### ISSUE TITLE
PLinq Aggregate - wrong computation

### ISSUE THREAD

**Issue by jl0pd** [dotnet/core#2730]
# Issue Title

Aggregate method from PLinq can't correctly collect results

# General

.Net Core: 2.2.203
Linux mint: 19.1
Linux kernel: 4.15.0-20-generic

When I try to find sum of some sequence using PLinq, I get wrong results.
Aggregate have problems when lambda is more complicated than ``x+y``. The more complex function it got, the less accurate result is.

For example, when fn(x) was ``return x``, Aggregate returned ``10``, as required;
When fn(x) was x ** 2, Aggregate returned 20;
x ** 3 -> 46
x ** 4 -> 116
While ``x`` all the time was just ``1``.

```C#
static void Test()
{
    Func<double, double> fn = x => x*x;
    double[] ar = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1}; // len is 10

    double realSum = ar.Aggregate((s, x) => s + fn(x));
    double parallelSum = ar.AsParallel().Aggregate((s, x) => s + fn(x));

    Console.WriteLine($"realSum: {realSum}");
    Console.WriteLine($"parallelSum: {parallelSum}");
}
```

**Comment by svick** [comment]
The problem with the code you posted is that since `ParallelEnumerable.Aggregate` performs the computation in parallel, it has to combine the results from different threads. And in this case, the only possible way to combine them is by invoking the provided function again, which will result in executing `fn` where it wouldn't be executed in `realSum`.

I believe the way to fix this is to use the more complex overload of `Aggregate`:

```c#
double parallelSum = ar.AsParallel().Aggregate(
    0d, (s, x) => s + fn(x), (s1, s2) => s1 + s2, s => s);
```

Though keep in mind that this does not behave exactly the same as `realSum`, because `realSum` does not call `fn` on the first value in the array, whereas the above `parallelSum` does.

In other words, this code:

```c#
Func<double, double> fn = x => x*x;
double[] ar = {2, 2, 2, 2, 2, 2};

double realSum = ar.Aggregate((s, x) => s + fn(x));
double badSum = ar.AsParallel().Aggregate((s, x) => s + fn(x));
double parallelSum = ar.AsParallel().Aggregate(
    0d, (s, x) => s + fn(x), (s1, s2) => s1 + s2, s => s);

Console.WriteLine($"realSum: {realSum}");
Console.WriteLine($"badSum: {badSum}");
Console.WriteLine($"parallelSum: {parallelSum}");
```

prints:

```
realSum: 22
badSum: 50
parallelSum: 24
```

**Comment by jl0pd** [comment]
Thanks, that helped!
